packages/teleterm/src/main.ts

```diff
@@ -13,118 +13,132 @@ import { ConfigServiceImpl } from 'teleterm/services/config';
 import { createFileStorage } from 'teleterm/services/fileStorage';
 import { WindowsManager } from 'teleterm/mainProcess/windowsManager';
 
-const settings = getRuntimeSettings();
-const logger = initMainLogger(settings);
-const fileStorage = createFileStorage({
-  filePath: path.join(settings.userDataDir, 'app_state.json'),
-});
-const configService = new ConfigServiceImpl();
-const windowsManager = new WindowsManager(fileStorage, settings);
-
-process.on('uncaughtException', error => {
-  logger.error('', error);
-  app.quit();
-});
-
-// init main process
-const mainProcess = MainProcess.create({
-  settings,
-  logger,
-  configService,
-  fileStorage,
-});
-
-app.on(
-  'certificate-error',
-  (event, webContents, url, error, certificate, callback) => {
-    // allow certs errors for localhost:8080
-    if (
-      settings.dev &&
-      new URL(url).host === 'localhost:8080' &&
-      error === 'net::ERR_CERT_AUTHORITY_INVALID'
-    ) {
-      event.preventDefault();
-      callback(true);
-    } else {
-      callback(false);
-      console.error(error);
-    }
-  }
-);
-
-app.on('will-quit', () => {
-  fileStorage.putAllSync();
-  globalShortcut.unregisterAll();
-  mainProcess.dispose();
-});
-
-app.whenReady().then(() => {
-  if (mainProcess.settings.dev) {
-    // allow restarts on F6
-    globalShortcut.register('F6', () => {
-      mainProcess.dispose();
-      const [bin, ...args] = process.argv;
-      const child = spawn(bin, args, {
-        env: process.env,
-        detached: true,
-        stdio: 'inherit',
-      });
-      child.unref();
-      app.quit();
-    });
-  }
-
-  windowsManager.createWindow();
-});
-
-// Limit navigation capabilities to reduce the attack surface.
-// See TEL-Q122-19 from "Teleport Core Testing Q1 2022" security audit.
-//
-// See also points 12, 13 and 14 from the Electron's security tutorial.
-// https://github.com/electron/electron/blob/v17.2.0/docs/tutorial/security.md#12-verify-webview-options-before-creation
-app.on('web-contents-created', (_, contents) => {
-  contents.on('will-navigate', (event, navigationUrl) => {
-    logger.warn(`Navigation to ${navigationUrl} blocked by 'will-navigate'`);
-    event.preventDefault();
+if (app.requestSingleInstanceLock()) {
+  initializeApp();
+} else {
+  // All windows will be closed immediately without asking the user,
+  // and the before-quit and will-quit events will not be emitted.
+  app.exit(1);
```

Comment:
>   // All windows will be closed immediately without asking the user,
>   // and the before-quit and will-quit events will not be emitted.

That's a very good choice here. 👍

Have you checked if we need an explicit `return` after `app.exit()`? The docs don't give a clear answer IMHO. I know that you need one after `app.quit()`, otherwise Electron just continues execution after `app.quit()`.

Technically everything's within `initializeApp` so there's nothing else to be executed after that `if`. But I'd add that `return` if it's needed, just to be explicit.

Reply:
nvm you cannot call `return` in a top-level if